### PR TITLE
feat(cell): Air780E (EC618) 4G LTE Cat 1 cellular modem support

### PR DESCRIPTION
**Pull request by ndoo**
## Summary

Adds the Air780E (EC618) as a new network interface for the MQTT client: an AT-command driver (`ATModem`/`EC618Modem`), a bearer/link-state machine (`cellBearer`), an Arduino `Client` for MQTT (`CellClient`) with TLS terminated by the modem itself via `AT+CIPSSL`, diagnostics (`cellDiag`) on a UI screen and as `CellularMetrics` telemetry, and a System Action toggle to power the modem on/off without a reboot. Config lives in `NetworkConfig` protobuf fields, mirroring `wifi_*`.

Named `cell*`/`HAS_CELLULAR` rather than `air780e*` so another EC618-family module or vendor can slot in later.

## What's included

- **Modem driver** (`src/mesh/cell/`): `ATModem`/`EC618Modem` (UART framing, AT dialect, TLS via `AT+CIPSSL`), `cellBearer` (link state machine), `CellClient` (single-socket IPv4 `Client`), `cellDiag`/`cellDiagParse` (signal/registration/battery sweep + parsers).
- **Network integration**: `HAS_CELLULAR` arm in `MQTT.{h,cpp}`'s client typedef, `isConnectedToNetwork()`, and `reconnect()`'s TLS branch; `CellularTelemetry` module sends `cellDiag` output as `Telemetry.cellular_metrics`.
- **Device UI**: Cellular System Action toggle and screen frame (`MenuHandler.{h,cpp}`).
- **Config/protobufs**: new `NetworkConfig` fields, `DeviceMetadata.hasCellular`, `Telemetry.CellularMetrics` (see Protobufs below).
- **Variants**: renamed T-Deck-Pro's `MODEM_POWER_EN`/`MODEM_PWRKEY` to this driver's `PIN_MODEM_*` naming - no behavior change.
- **Tests**: `test/test_cell_diag/`, 21 cases over the diagnostics parsers.

## Known limitations

- No listening sockets (`AT+CIPSTART` is outbound-only) - phone API server, web server, UDP multicast excluded.
- IPv4 only on the socket API.
- Power cycling an unresponsive modem requires `PIN_MODEM_EN` wired.

## Needs testing

- SIM hot-insert detection (`MODEM_SIM_HOTPLUG`) - bench module has no `USIM_CD` wired.

## PWRKEY / EN wiring

| PWRKEY | EN  | Power-on                 | Power-off                                     | SIM re-detect   | Recovers a hung modem? |
| ------ | --- | ------------------------ | ---------------------------------------------- | ---------------- | ------------------------ |
| No     | No  | Auto-boot only           | Not possible                                  | `AT+RESET` only | No                     |
| Yes    | No  | PWRKEY pulse             | Soft shutdown                                 | `AT+RESET`      | No                     |
| No     | Yes | EN assert, auto-boot     | EN de-assert (regulator off)                  | EN cycle        | Yes                    |
| Yes    | Yes | EN assert → PWRKEY pulse | Soft shutdown -> EN de-assert (regulator off) | EN cycle        | Yes - recommended      |

## Troubleshooting

Frequent hangs or resets are likely brownout (caused by multiple back-to-back TX bursts sagging the supply) - rule out power stability before suspecting a modem firmware bug:

- Confirm headroom for ~2A TX spikes, not just ~1A sustained.
- Use a high-current source (e.g. a LiPo cell) on short wires, or add bulk capacitance near the modem.
- Wire both PWRKEY and EN if unsure - only that combination self-recovers from a hang.

**VIN vs VBAT:** VBAT is 3.3-4.3V; the VIN pin is regulated from a wider range (check product info for voltage range), and that regulator is controlled by the `EN` pin.

## Build-time configuration

`build_flags` only - no in-tree board ships this modem yet. Required: `HAS_CELLULAR=1`, `USE_EC618`, `PIN_MODEM_TX`/`PIN_MODEM_RX`. `PIN_MODEM_PWRKEY`/`PIN_MODEM_EN` optional (see wiring table). Full macro list and defaults in `ATModem.h`/`EC618Modem.h`.

## Protobufs

From [meshtastic/protobufs#1030](https://github.com/meshtastic/protobufs/pull/1030) (not yet merged): `NetworkConfig.cell_enabled`/`cell_apn`/`cell_apn_user`/`cell_apn_pass`/`cell_roaming_enabled`, `DeviceMetadata.hasCellular`, `Telemetry.cellular_metrics` (new `CellularMetrics` message: rsrp/rsrq/band/access_tech/modem_voltage_mv/operator_name, all proto3-optional), and `TelemetryConfig.cellular_measurement_enabled`/`cellular_update_interval`.

Until #1030 merges, the submodule is pinned to the fork's commit instead (same technique as [meshtastic/firmware#10755](https://github.com/meshtastic/firmware/pull/10755)) - `.gitmodules` unchanged, forks share object storage with upstream. No extra setup needed to check out this PR.

## Testing

Bring-up used a stock Heltec V3 with the modem wired to its exposed header (no in-tree board ships this modem). Append to `variants/esp32s3/heltec_v3/platformio.ini` to reproduce:

```ini
; Bring-up setup for the Air780E modem - not a real variant. Serial2 keeps
; Serial1 free for GPS. See the PR description for wiring/config details.
; Wire a high-current 5V or 4.2V source (e.g. a power-path-switched VBUS) to
; the module's VIN/5V input. Check its datasheet for VIN range - see PR notes.
[env:heltec-v3-air780e]
extends = env:heltec-v3
board_level = extra
build_flags =
  ${env:heltec-v3.build_flags}
  -DHAS_CELLULAR=1
  -DUSE_EC618
  -DPIN_MODEM_EN=4
  -DPIN_MODEM_PWRKEY=5
  -DPIN_MODEM_RX=6
  -DPIN_MODEM_TX=7
  -DMODEM_SERIAL=Serial2
```

Done:

- `heltec-v3-air780e` (`HAS_CELLULAR=1`) and `heltec-v3` (`HAS_CELLULAR=0`) both build-verify clean.
- `t-deck-pro`, `t-deck-pro-v1_1`, `tbeam`: build-verified only.
- Native suite in Docker: 864/866 pass; the 2 failures (`test_packet_signing` `B11`/`B12`) predate this branch.
- Bench setup: modem powers on, attaches, gets an IP, publishes MQTT with and without TLS; diagnostics sweep confirmed on-screen; System Action toggle verified powering the modem off and back on.

## 🤝 Attestations

- [x] I have tested that my proposed changes behave as described.
- [ ] I have tested that my proposed changes do not cause any obvious regressions on the following devices:
  - [x] Heltec (Lora32) V3 - stock `heltec-v3` and the `heltec-v3-air780e` bring-up setup, both on hardware
  - [ ] LilyGo T-Deck
  - [ ] LilyGo T-Beam - build-verified only
  - [ ] RAK WisBlock 4631
  - [ ] Seeed Studio T-1000E tracker card
  - [x] Other (please specify below): LilyGo T-Deck-Pro / T-Deck-Pro v1.1 - build-verified only (renamed `PIN_MODEM_*` macros, no behavior change)

## Summary by CodeRabbit

* **New Features**
  * Added cellular modem connectivity with power control, roaming, sockets, TLS, and MQTT transport.
  * Added support for EC618-based modems.
  * Added a Cellular status screen showing signal, network, IP, operator, band, and voltage.
  * Added cellular diagnostics and optional telemetry sharing over mesh or phone connections.
  * Added automatic fallback to cellular when Wi-Fi or Ethernet is unavailable.

* **Bug Fixes**
  * Improved syslog support for applicable network transports.
  * Improved modem power-control configuration across supported devices.

* **Tests**
  * Added coverage for cellular diagnostic response parsing and edge cases.

### CODE REVIEW

**Comment by ndoo** [comment]
@coderabbitai review

File: src/mesh/cell/ATModem.cpp [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+    queue.clear();
+    commandActive = false;
+    awaitingFinalUrc = false;
+    active = Command();
+    response = "";
+    rxLine = "";
+    model = "";
+    version = "";
+    pulseToOff = false;
+}
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/ATModem.cpp:95]: Fixed in b51b63f44 - `clearSession()` now drains the queue before `cancelActive()` and notifies each dropped command's callback with `ATResult::Error`, matching the callback-runs-once contract.

File: src/mesh/cell/ATModem.cpp [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+    cancelActive();
+    queue.clear();
+    commandActive = false;
+    awaitingFinalUrc = false;
+    active = Command();
+    response = "";
+    rxLine = "";
+    model = "";
+    version = "";
+    pulseToOff = false;
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/ATModem.cpp:90]: Fixed in b51b63f44 - `clearSession()` now resets `binaryRemaining`/`binarySink`/`binaryBuf` alongside `rxLine`.

File: src/mesh/cell/cellBearer.cpp [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+            initCellModem();
+        cellLastResponse = millis();
+        cellFailCount = 0;
+        simAbsentLogged = false;
+        simResetCount = 0;
+        setCellState(CELL_BOOTING);
+    } else if (cellModem) {
+        cellModem->powerOff();
+        setCellState(CELL_OFF);
+    }
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/cellBearer.cpp:350]: Fixed in 0651d9627 - `setCellState()` now rejects any transition other than `CELL_OFF` while `cellEnabled` is false, so a stale queued callback can't revive the state machine after disable.

File: src/mesh/cell/cellBearer.cpp [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+            backoff = CELL_BACKOFF_MAX_MS;
+        cellLastResponse = now;
+        setCellState(CELL_REG_WAIT);
+        return backoff;
+    }
+
+    case CELL_SIM_WAIT:
+        // With USIM_CD wired the module reports insertion itself, so waiting
+        // costs nothing and a reset would only throw away a working session.
+        if (!cellModem->hasSimHotplug() && simAbsent && (uint32_t)(now - simWaitSince) >= simResetBackoff())
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/cellBearer.cpp]: Fixed in 0651d9627 - both elapsed-time checks (`cellLastResponse` silence, `simWaitSince` backoff) now use `Throttle::isWithinTimespanMs()`.

File: src/mesh/cell/CellClient.cpp [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+{
+    uint32_t deadline = millis() + timeoutMs;
+    while (!done()) {
+        if ((int32_t)(millis() - deadline) >= 0)
+            return done();
+        cellModem->service();
+        delay(1);
+    }
+    return true;
+}
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/CellClient.cpp:90]: Not changed. `CellClient::waitFor()`'s `cellModem->service(); delay(1);` loop is the same watchdog-safe blocking idiom used elsewhere in this codebase for bounded waits (`NodeDB.cpp`, `mesh/eth/ethOTA.cpp`, `mesh/eth/ethApiHandlers.cpp`) - `delay(1)` yields to FreeRTOS, which services the idle-task watchdog on ESP32. There's no `esp_task_wdt`/task-level feed call in this codebase to add here.

File: src/mesh/cell/cellDiagParse.cpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+    long rsrq = intField(line, 4);
+    long rsrp = intField(line, 5);
+    if (rsrq < 0 || rsrp < 0 || rsrq == 255 || rsrp == 255)
+        return false;
+
+    if (rsrpDbm)
+        *rsrpDbm = (int16_t)(-140 + rsrp);
+    if (rsrqDb)
+        *rsrqDb = -20.0f + 0.5f * (float)rsrq;
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/cellDiagParse.cpp:103]: Fixed in 48f51dcd6 - `rsrq` is now rejected outside 0-34 and `rsrp` outside 0-97, not just negative/255. Added a regression test for out-of-range values.

File: src/mesh/cell/EC618Modem.cpp [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        break;
+
+    case BEARER_CSTT: {
+        // CSTT is not optional even with no APN: it moves the module from IP INITIAL to
+        // IP START, and CIICR is only valid from IP START. An empty APN asks for the subscription default.
+        String cmd = String("AT+CSTT=\"") + config.network.cell_apn + "\",\"" + config.network.cell_apn_user + "\",\"" +
+                     config.network.cell_apn_pass + "\"";
+        submit(cmd.c_str(), 10000, cb);
+        break;
+    }
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/EC618Modem.cpp:132]: Fixed in a69f788ac - added `sanitizeAtArg()`, applied to `cell_apn`/`cell_apn_user`/`cell_apn_pass` in `bringUpBearer()` and to `host` in `socketOpen()`.

File: src/mesh/cell/EC618Modem.cpp [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    int secondComma = s.indexOf(',', firstComma + 1);
+
+    int len = s.substring(firstComma + 1, secondComma < 0 ? s.length() : secondComma).toInt();
+    int remaining = secondComma < 0 ? 0 : s.substring(secondComma + 1).toInt();
+    socketEvents.onRxPending(remaining > 0);
+
+    if (len > 0) {
+        auto onData = socketEvents.onData;
+        captureBinary(len, [onData](const uint8_t *data, size_t n) { onData(data, n); });
+    }
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/EC618Modem.cpp:186]: Fixed in a69f788ac - `len` is now clamped to `socketRecvMax()` before capture, and a `captureBinary()` rejection is logged.

File: src/mesh/cell/EC618Modem.cpp [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+// The module's own power-on banner - seeing it while a command is in flight means the
+// modem reset itself mid-command, so fail that command now rather than wait out its
+// full timeout. Text verified on Air780E/EX only; a sibling EC618 module should confirm
+// it prints the same lines before relying on this.
+void EC618Modem::onUnexpectedReboot()
+{
+    LOG_WARN("Modem rebooted unexpectedly mid-command");
+    cancelActive();
+}
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/mesh/cell/EC618Modem.cpp:199]: Fixed in a69f788ac - `onUnexpectedReboot()` now calls `expectReboot()` instead of `cancelActive()`, so `sessionInit()` reruns once AT answers again.

File: src/modules/Modules.cpp [hunk 10]
```diff
@@ -66,6 +66,9 @@
 #if HAS_TELEMETRY
 #include "modules/Telemetry/DeviceTelemetry.h"
 #endif
+#if HAS_TELEMETRY && HAS_CELLULAR
+#include "modules/Telemetry/CellularTelemetry.h"
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/modules/Modules.cpp]: Fixed in 29f1a7c5c - both the include and the module construction are now gated behind `!MESHTASTIC_EXCLUDE_CELLULAR_TELEMETRY` as well.

File: src/modules/Telemetry/CellularTelemetry.cpp [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        m->variant.cellular_metrics.rsrq = diag.rsrq;
+    }
+    if (diag.bandValid) {
+        m->variant.cellular_metrics.has_band = true;
+        m->variant.cellular_metrics.band = diag.band;
+        m->variant.cellular_metrics.has_access_tech = true;
+        m->variant.cellular_metrics.access_tech = diag.act;
+        m->variant.cellular_metrics.has_operator_name = true;
+        strncpy(m->variant.cellular_metrics.operator_name, diag.operatorName, sizeof(m->variant.cellular_metrics.operator_name));
+    }
```
[the remark this answers is not in the document]
Comment (ndoo) [reply src/modules/Telemetry/CellularTelemetry.cpp:99]: Fixed in 29f1a7c5c - copies `sizeof(operator_name) - 1` bytes and explicitly terminates the last byte.

**Comment by ndoo** [comment]
Addressed the review findings as 8 commits, grouped functionally:

- b51b63f44 fix(cell): harden ATModem session teardown and error handling
- a69f788ac fix(cell): validate and sanitize EC618 modem I/O
- 0651d9627 fix(cell): guard bearer state after disable, use Throttle for timing
- 58b3dcf20 fix(cell): release the modem command slot on CellClient timeout paths
- 48f51dcd6 fix(cell): validate CESQ signal range, use Throttle for diag sweep
- 29f1a7c5c fix(cell): NUL-terminate operator_name and gate telemetry module
- 13bc8d07a fix(cell): gate the MQTT reachability probe at runtime
- 260dc6b82 fix(cell): implement the cellular frame visibility toggle

One finding (`CellClient::waitFor()`'s blocking loop) was left as-is - its `service(); delay(1);` pattern already matches the watchdog-safe blocking idiom used elsewhere in this codebase (`NodeDB.cpp`, `mesh/eth/ethOTA.cpp`, `mesh/eth/ethApiHandlers.cpp`); see the inline reply for detail.

Native test suite: 867 cases, 2 pre-existing failures unrelated to this change (confirmed by reproducing them with these commits reverted). Hardware: build-flashed to the heltec-v3-air780e bring-up rig and confirmed live - cellular MQTT is receiving and publishing real mesh traffic (CIPRXGET/CIPSEND/CEREG all succeeding) at CELL_IP_UP.

@coderabbitai review

**Comment by ndoo** [comment]
Addressed the re-review findings in 539d54536:

- Duplicate (`EC618Modem.cpp` `onCipRxGet()`): a rejected `+CIPRXGET: 2` payload (oversized, or a capture already in flight) returned without consuming the announced bytes, so `pumpSerial()` would parse them as AT line text. Added `ATModem::discardBinary()` - a bounded diversion that drops bytes without buffering them - and use it for an oversized length; a `captureBinary()` rejection (the one diversion slot already occupied) now calls `expectReboot()` to resync instead of only logging.
- Nitpick: shortened `sanitizeAtArg()`'s comment to the repo's one-to-two-line convention.

Build-verified (`heltec-v3-air780e`) and `test_cell_diag` (21/21) both pass.

@coderabbitai review

**Comment by ndoo** [comment]
Working on 2 things,

1. Reduce the changes to MQTT.cpp
2. The module does actually support SSL, ~but not all module firmware may support it; adding runtime check but also the existing Wi-Fi MQTT code is not designed for runtime capability check for SSL.~ implementing it forced via a macro mirroring Wi-Fi to reduce the changes to MQTT.cpp

**Comment by ndoo** [comment]
- Added TLS support for cellular MQTT via the modem's internal SSL pathway, `AT+CIPSSL`.
- Hardware-tested on the test variant (see PR):
  - TLS connects to `mqtt.meshtastic.org:8883`
  - A bad handshake intentionally fails without a crash rather than downgrading to non-TLS
  - Non-TLS tested to still work (without regression)
- Updated PR text to simplify

@coderabbitai review

**Comment by ndoo** [comment]
Trying to reduce the MQTT.cpp changes further but hitting a roadblock because the only way to not rework how the MQTT client is defined, is if we maintain the limitation today with Wi-Fi and Ethernet being mutually exclusive.

In the case of forcing a user to pick between Wi-Fi, Ethernet and 4G variants, MQTT.cpp changes would be far less since we don't need to be careful of which client ends up being used for MQTT.

**Comment by ndoo** [comment]
- Reduced the MQTT.cpp transport/TLS selection code further:
  - Add a `cellClient()` accessor for picking Wi-Fi/Ethernet (if enabled in preference) --> cellular (if enabled in preference)
  - Used it to simplify `reconnect()`'s TLS branch - one `setTlsEnabled()` call instead of two
  - Merged `isConnectedToNetwork()`/`activeClient()`'s duplicated transport-preference chain into one shared helper
- Renamed `HAS_SYSLOG` to `HAS_NETWORKING_UDP` to better reflect why syslog was gated
- Above changes made to generally reduce impact to existing MQTT.cpp
- Added docstrings across cellular driver headers (flagged by @coderabbitai's docstring coverage check)
- Verified:
  - Build-verified (`heltec-v3-air780e`, `heltec-v3`).
  - Native suite unchanged (867 cases, same 2 pre-existing failures).
  - Hardware-tested on the `heltec-v3-air780e` bring-up env - TLS cellular MQTT still connects and receives real downlink traffic.

@coderabbitai review

**Comment by ndoo** [comment]
- Fixed in `ace02b4c3`: wait for the `AT+CIPSSL` result before opening the socket, refuse the connect if TLS was requested and it wasn't `Ok`.
- Mirrors the existing `socketRxMode()` wait-then-check pattern a few lines above.
- Build-verified and hardware-tested - TLS cellular MQTT still connects with no added latency.

@coderabbitai review

**Comment by ndoo** [comment]
@coderabbitai The docstring coverage failure doesn't seem to have changed after pushing the docstrings in `ae65f4a2a docs(cell): document remaining public API gaps`?

**Comment by ndoo** [comment]
- Fixed in `142ff484c`: `isValidConfig()` now checks `activeTransport() == NetTransport::CELLULAR` instead of `isCellularAvailable()`, so the reachability probe is only skipped when cellular is actually the selected transport, not merely present alongside WiFi/Ethernet.
- Folds in the existing `isConnectedToNetwork()` check and trims the adjacent comment to two lines.
- Build-verified (`heltec-v3-air780e`) and native test-verified - 867 cases, all passing except two pre-existing `test_packet_signing` PKI failures unrelated to this change and reproduced on the tree without it.

@coderabbitai resume

**Comment by ndoo** [comment]
@coderabbitai review

**Comment by ndoo** [comment]
@coderabbitai the docstring pre-merge check seems stuck

**Comment by ndoo** [comment]
Closing this myself because I don't think anyone particularly wants this for 2.8.

But it's still going to be good reference for later anyway, so leaving this here in the annals of PR history.
